### PR TITLE
Volunteer Certificate Download Button - Full state flow

### PR DESCRIPTION
**Pull request by mendelB**
### What's this PR do?

This pull request updates the `CertificateDownloadButton` to include the complete experience across success, loading, and error states.

### How should this be reviewed?
Hopefully, the comments help! I also hope this intimidating component actually proves to be deceptively simple.

Do you agree with my rationale outlined below?

Does the Error state make sense here? (Rendering an `ErrorBlock` in place of the button.

The review app is failing to build or release upon various tries. I _think_ it's a Heroku issue 🤔 
GIF of the download flow: https://github.com/DoSomething/phoenix-next/pull/2067#issuecomment-620666574

### Any background context you want to provide?
Some context per #2063:

We opted to include a more imperative PDF rendering / download logic flow so that we have more control of the logic, and error handling than the simple download link the library provides.

We wanted to avoid using a phantom `<a>` tag to download the PDF since that increases complexity.

Our compromise was manipulating an existing `<a>` tag 'download button'.

I found while attempting to introduce error and loading state into the component, that the complexity inherently grew to the extent that it felt OK to use the 'phantom' link to avoid a dizzying amount of DOM manipulation to keep an `<a>` tag rendering properly in the DOM across the various states.

In summary: my assessment is that the complexity inherent in the endeavor of a (well relatively) rich PDF downloading experience necessitates a complex component.

That being said I tried to break things down as elegantly as I could! But you'll be the judge of that, oh esteemed reviewer.

### Relevant tickets

References [Pivotal #172439408](https://www.pivotaltracker.com/story/show/172439408).

### Checklist

- [ ] This PR has been added to the relevant Pivotal card.
- [ ] Documentation added for new features/changed endpoints.
- [ ] Added screenshots of front-end changes on small, medium, and large screens.
- [ ] Added appropriate feature/unit tests.

### CODE REVIEW

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 1]
```diff
@@ -1,55 +1,124 @@
...
+  const [pdfLink, setPdfLink] = useState(null);
 
-  const handleClick = () => {
-    if (!hasGeneratedPdf) {
+  /**
+   * Handle PDF rendering errors.
+   *
+   * @param {Error} err
+   */
+  const handleError = err => {
```
Comment (mendelB) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:20]: You'll notice the awkward `err` parameter (here and below). My poor 🧠could not come up with a better name for the hook `error` and thus we selfishly claim `error` in the upper scope. Any ideas?

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 2]
```diff
@@ -1,55 +1,124 @@
...
+   *
+   * @return {Promise|Undefined}
+   */
+  const generatePdfLink = () =>
+    new Promise(resolve => {
+      // We'll be 'loading' until the PDF link is generated.
+      setLoading(true);
+
+      // Playing it safe to catch any errors caused by the initial PDF rendering step.
       try {
```
Comment (mendelB) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:40]: (This isn't new to this PR, but I forgot to explain): Frustratingly, there doesn't seem to be an in-house solution for safely rendering the PDF, hence the `try` (The [in-house pdf link](https://react-pdf.org/advanced#on-the-fly-rendering) does seem to catch and log errors to the console, so you could argue that it's a win to take that over ourselves so we can report these.)

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 3]
```diff
@@ -1,55 +1,124 @@
...
+  const buttonText = () => {
+    if (certificatePost.pending) {
+      return 'Pending';
+    }
+
+    if (loading) {
+      return 'Loading';
     }
+
+    return 'Download';
```
Comment (mendelB) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:102]: I know it's just plain text, but across three potential states, the usual ternary did not seem like the way to go.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 4]
```diff
@@ -1,55 +1,124 @@
...
 
-  if (certificatePost.pending) {
-    return (
-      <button type="button" disabled className="btn w-full py-4 text-lg">
-        Pending
-      </button>
-    );
+  if (error) {
+    return <ErrorBlock error={error} />;
   }
```
Comment (mendelB) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:107]: This does stretch the table cell a bit. But it didn't seem terrible enough not to employ the convenience of our in-house block -- at least for launch! Lemme know if you disagree tho!

image

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 5]
```diff
@@ -1,55 +1,124 @@
...
+
+  /**
+   * Generate the certificate PDF data, and build a phantom download link.
+   *
+   * @return {Promise|Undefined}
+   */
+  const generatePdfLink = () =>
+    new Promise(resolve => {
+      // We'll be 'loading' until the PDF link is generated.
+      setLoading(true);
```
Comment (mendelB) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:37]: It might make sense to move the loading logic back into the `handleClick` below so that the PDF generation logic is more encapsulated here? 🤔

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 6]
```diff
@@ -1,55 +1,124 @@
...
+    console.error(err);
+  };
+
+  /**
+   * Generate the certificate PDF data, and build a phantom download link.
+   *
+   * @return {Promise|Undefined}
+   */
+  const generatePdfLink = () =>
+    new Promise(resolve => {
```
Comment (mendelB) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:35]: We use a Promise here so that we can asynchronously 'click' the button from the `handleClick`, I initially had the click happen within this method, but it did feel better to keep it outside, keeping the job of this function to merely resolve the link. (We could even make it a more conventional resolve/reject pattern based on the `catch`es, and move the error handling to the `handleClick` as well per https://github.com/DoSomething/phoenix-next/pull/2067/files#r416634482 🤔)

**Comment by weerd** [comment]
Appreciate the comments, but I don't know if it's helpful to have this many... lol. Finding a few of a comments more distracting than helpful, since there are a lot of them 😅

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 7]
```diff
@@ -1,55 +1,124 @@
...
+      // We'll be 'loading' until the PDF link is generated.
+      setLoading(true);
+
+      // Playing it safe to catch any errors caused by the initial PDF rendering step.
       try {
+        // Render the PDF template.
         pdf(<CertificateTemplate certificatePost={certificatePost} />)
+          // Convert the data to a Blob object.
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:43]: Remove this comment. It's straightforward that it's converting to a blob via `toBlob()` :)

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 4 above)
Comment (weerd) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:107]: Not ideal, but ya, seems fine for now. I wonder if we need a mini version of the ErrorBlock... ErrorArtifact? haha

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 8]
```diff
@@ -1,55 +1,124 @@
...
+        'is-loading': loading,
+      })}
+      disabled={certificatePost.pending || loading}
+      // If a sneaky user removes the 'disabled' attribute we don't want to trigger the PDF download.
+      onClick={() => !certificatePost.pending && handleClick()}
     >
-      Download
-    </a>
+      {buttonText()}
+    </button>
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:121]: Could we use an `ElementButton` here? And setting it in the flow during the pending, loading, ready process.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 9]
```diff
@@ -1,55 +1,124 @@
...
+        'bg-blue-500 hover:bg-blue-300': !certificatePost.pending,
+        'is-loading': loading,
+      })}
+      disabled={certificatePost.pending || loading}
+      // If a sneaky user removes the 'disabled' attribute we don't want to trigger the PDF download.
+      onClick={() => !certificatePost.pending && handleClick()}
     >
-      Download
-    </a>
+      {buttonText()}
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:120]: Would using a state property be more appropriate here? and then setting it via part of the flow in code when hitting pending, loading, ready, etc?

**Comment by weerd** [comment]
@mendelB just so I understand the flow correctly, when I load the page with the table, do the PDFs start generating after I load the table and then they each change to a download button as soon as the PDF is ready? Or do I as a user need to interact with the button to signal I want a PDF to generate?

**Comment by mendelB** [comment]
> @mendelB just so I understand the flow correctly, when I load the page with the table, do the PDFs start generating after I load the table and then they each change to a download button as soon as the PDF is ready? Or do I as a user need to interact with the button to signal I want a PDF to generate?

Ah flip, I meant to link to the review app, but the deploy kept failing. and then forgot to attach a GIF. It's the latter! The user clicks the download button, the PDF generates (on the first click), and then downloads. I'll attach a GIF!

**Comment by weerd** [comment]
> Ah flip, I meant to link to the review app, but the deploy kept failing. and then forgot to attach a GIF. It's the latter! The user clicks the download button, the PDF generates (on the first click), and then downloads. I'll attach a GIF!

Ah, ya, ok! I was confused cause I was using the image you added to one of the comments and there all the buttons say "Pending" so got the impression they for some reason were generating on their own and initially load as "Pending" lol. GIF in discussion would be helpful to represent the flow visually as reference! 👍

**Comment by mendelB** [comment]
@weerd (wow KAP does _not_ capture the `cursor-wait` spinner properly )

Kapture 2020-04-28 at 11 07 03

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 10]
```diff
@@ -1,55 +1,124 @@
...
+          // Convert the data to a Blob object.
           .toBlob()
           .then(blob => {
-            pdfLink.current.href = URL.createObjectURL(blob);
-            pdfLink.current.click();
+            // Create a 'phantom' link so we can download the PDF on the user's behalf.
+            const phantomLink = document.createElement('a');
 
-            setHasgeneratedPdf(true);
+            // Convert the PDF Blob to a URL and assign it as the link's href.
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:49]: Remove this comment. Code explains it.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 11]
```diff
@@ -1,55 +1,124 @@
...
+            // Clicking the link will now download the pdf automatically.
+            phantomLink.href = URL.createObjectURL(blob);
+            // Assign the PDF filename.
+            // @TODO: Update filename per https://www.pivotaltracker.com/story/show/172439408/comments/213717690.
+            phantomLink.download = 'certificate.pdf';
+
+            // Save this download link in state so we can just 'click' it again if necessary.
+            setPdfLink(phantomLink);
+
+            // Cool, we're no longer loading.
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:59]: Remove comment.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 12]
```diff
@@ -1,55 +1,124 @@
...
   return (
-    // eslint-disable-next-line jsx-a11y/anchor-is-valid
-    <a
-      role="link"
-      tabIndex="0"
-      download="certa.pdf"
-      ref={pdfLink}
-      className="btn w-full py-4 text-lg bg-blue-500 hover:bg-blue-300"
-      onClick={handleClick}
+    <button
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:110]: Curious what kept you from potentially using the `ElementButton` here? The "loading" state? We could add that to the component similar to how it's set up in the `ToggleButton`. Since we have the approval to move to blurple buttons, the design having these as the old blue should not be the reason to not be using the new buttons :)

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 3 above)
Comment (weerd) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:102]: See suggestion in [comment](https://github.com/DoSomething/phoenix-next/pull/2067/files#r416684693) below.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 13]
```diff
@@ -1,55 +1,124 @@
...
           .toBlob()
           .then(blob => {
-            pdfLink.current.href = URL.createObjectURL(blob);
-            pdfLink.current.click();
+            // Create a 'phantom' link so we can download the PDF on the user's behalf.
+            const phantomLink = document.createElement('a');
 
-            setHasgeneratedPdf(true);
+            // Convert the PDF Blob to a URL and assign it as the link's href.
+            // Clicking the link will now download the pdf automatically.
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:50]: Probs don't need this comment either; found it a bit confusing since the "click" isn't specified in this part of the code but in the `handleClick`.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 14]
```diff
@@ -1,55 +1,124 @@
...
+   */
+  const handleClick = () => {
+    // If we've already generated a pdfLink, we can simply 'click' it to download the pdf.
+    if (pdfLink) {
+      pdfLink.click();
+      return;
+    }
+
+    // generate the phantom PDF link, and 'click' it to download the pdf!
+    generatePdfLink().then(link => link.click());
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:86]: Noice! 👯

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 5 above)
Comment (weerd) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:37]: Ya, I think that makes sense to me.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 4 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:107]: Ooh yeah lol that would actually be interesting! We do have other buttons now that account for pretty dynamic state like the [Cause Preference](https://github.com/DoSomething/phoenix-next/blob/f3b56129b13ebba7e2b877b06b64c2f0e7bf61db/resources/assets/components/pages/AccountPage/Interests/CausePreferenceItem.js#L46-L48) and [Email Subscription](https://github.com/DoSomething/phoenix-next/blob/f3b56129b13ebba7e2b877b06b64c2f0e7bf61db/resources/assets/components/pages/AccountPage/Subscriptions/EmailSubscriptionItem.js#L44-L46) ones!

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 9 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:120]: That's an interesting idea! I'll give it whirl.

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 12 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:110]: No special reason! I had a @TODO to try and implement one of your new ones, but it got lost in the shuffle. I'll try the `ElementButton`!

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 5 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js]: Done! In https://github.com/DoSomething/phoenix-next/pull/2067/commits/fec8dde03ced081e14cdc131e10abee0058458ff

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 6 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js]: Better organized in https://github.com/DoSomething/phoenix-next/pull/2067#discussion_r416636108

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 9 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js]: This was a great idea! I ended up using `useEffect` as well to toggle 'loading' on and off while the `loading` state was updated and I think it looks pretty neat now!
https://github.com/DoSomething/phoenix-next/pull/2067/commits/c406206ede0c5e83021db26522402391f420b9e8

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 15]
```diff
@@ -1,55 +1,110 @@
...
 
 const CertificateDownloadButton = ({ certificatePost }) => {
-  const [hasGeneratedPdf, setHasgeneratedPdf] = useState(false);
-  const pdfLink = useRef(null);
+  const [error, setError] = useState(null);
+  const [loading, setLoading] = useState(false);
+  const [pdfLink, setPdfLink] = useState(null);
+  const [buttonText, setButtonText] = useState(
+    certificatePost.pending ? 'pending' : 'download',
+  );
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:16]: Noice! 👌

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 12 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js]: Added in https://github.com/DoSomething/phoenix-next/pull/2067/commits/ba9aefbe475b35b0580d7e5bec1d7f32e03e1f61

File: resources/assets/components/utilities/Button/ElementButton.js [hunk 16]
```diff
@@ -14,13 +14,14 @@ const ElementButton = ({
...
   isDisabled,
+  isLoading,
   onClick,
   text,
   type,
 }) => {
   return (
     <button
-      className={classnames('btn', className)}
+      className={classnames('btn', className, { 'is-loading': isLoading })}
```
Comment (weerd) [inline resources/assets/components/utilities/Button/ElementButton.js:24]: 💯

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 17]
```diff
@@ -1,55 +1,116 @@
...
-      ref={pdfLink}
-      className="btn w-full py-4 text-lg bg-blue-500 hover:bg-blue-300"
-      onClick={handleClick}
-    >
-      Download
-    </a>
+    <ElementButton
+      type="button"
+      className={classNames('w-full py-4 text-lg capitalize', {
+        'bg-blue-500 hover:bg-blue-300': !certificatePost.pending,
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:106]: We got approval to use blurple. Are you hesitant? haha 😏

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 1 above)
Comment (weerd) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:33]: Haha, ya. I think this is fine. As an alternative, I sometimes end up prefixing it based on what it's for to help differentiate and deal w/ variable name collision.

`generatePdfError`, `pdfLinkError`, etc. 🤷‍♂️

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 17 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js]: lololol updated in https://github.com/DoSomething/phoenix-next/pull/2067/commits/a6b6c034435c5cf8d2e787bb4c7223a4da8c0242

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js (hunk 1 above)
Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:33]: Ooh that's helpful! I'll update the naming before merging!

**Comment by mendelB** [comment]
Thanks for the feedback @weerd!! I feel a lot better about this component now!

File: resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js [hunk 18]
```diff
@@ -1,55 +1,116 @@
...
+      type="button"
+      className={classNames('w-full py-4 text-lg capitalize', {
+        'bg-blurple-500 hover:bg-blurple-300': !certificatePost.pending,
+      })}
+      isDisabled={certificatePost.pending || loading}
+      isLoading={loading}
+      // If a sneaky user removes the 'disabled' attribute we don't want to trigger the PDF download.
+      onClick={() => !certificatePost.pending && handleClick()}
+      text={buttonText}
+    />
```
Comment (weerd) [inline resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:113]: Oh just realized... but if it's blurple now, we could just use a `PrimaryButton` :) Not necessary for merge, but a thought for later, haha! But then would need to also update `PrimaryButton` with `isLoading`. Defs feels like a future update, but no worries on it for now.

Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:113]: Oh, snap! That'd be a nice follow-up refactor to remove some of the class efforts on our end here 😌

Comment (mendelB) [reply resources/assets/components/pages/AccountPage/Credits/CertificateDownloadButton.js:113]: Updated in https://github.com/DoSomething/phoenix-next/pull/2071/commits/eddd5e5ff7884730bb996f27bc1d9a45264e4d56!
